### PR TITLE
HDDS-5905. Race condition of deletion service and active object deletion

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

1. delete key is appended with objectId to make it unique 
2. Removed caching of deleteTable as its not used again due to uniqueness
3. and removed getting from deleteTable for old key (old key will not be present as delete key is unique)
    This is to add repeated info and overwrite the same key again in deleteTable, which can cause missing cleanup of latest block appended as there KeyDeletingService has already run in parallel.

Uniqueness for multiple flows:
- openKey cleanup: openKey appended with objectId
- key commit(unused block and old version): openKey appended with objectId
- delete key/keys: dbKey appended with objectId
- directoryPurge: key appended with objectId
- multipart upload abort: openKey with each part objectId
- multipart upload commit: openKey of part with old version objectId
- multipart upload complete: openKey with objectId of key

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5905

## How was this patch tested?

Unit test and integration test for delete behavior impact

### CODE REVIEW

**Comment by kerneltime** [comment]
cc @swamirishi @hemantk-12 @aswinshakil can you'll please take a look?

**Comment by szetszwo** [comment]
@sumitagrawl , we need to think about compatibility and upgrade.  The key format of `deletedTable` is changed.  What if there are existing entires in it?  How to upgrade from old format to new format?

**Review by smengcl — CHANGES_REQUESTED** [review]
Thanks @sumitagrawl for starting this PR.

Though unfortunately changing deletedTable key without fixing relevant Ozone snapshot logic will incur data loss with our current snapshot implementation. cc @aswinshakil 

Extra key conversion logic would have to be added for snapshot to function properly.

cc @prashantpogde

**Comment by sumitagrawl** [comment]
> Thanks @sumitagrawl for starting this PR.
> 
> Though unfortunately changing deletedTable key without fixing relevant Ozone snapshot logic will incur data loss with our current snapshot implementation. cc @aswinshakil
> 
> Extra key conversion logic would have to be added for snapshot to function properly.
> 
> cc @prashantpogde

Without making key unique, I am not getting any solution. Similar approach is followed for directory deletion also.
@smengcl @prashantpogde @aswinshakil Please check for solution for the current problem or do need change snapshot also for this? I am not getting coupling of snapshot with this key format as key is not unique itself.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMDirectoriesPurgeResponseWithFSO.java [hunk 1]
```diff
@@ -130,11 +130,10 @@ public void processPaths(OMMetadataManager omMetadataManager,
         }
 
         RepeatedOmKeyInfo repeatedOmKeyInfo = OmUtils.prepareKeyForDelete(
-            keyInfo, null, keyInfo.getUpdateID(), isRatisEnabled);
+            keyInfo, keyInfo.getUpdateID(), isRatisEnabled);
 
         String deletedKey = omMetadataManager
-            .getOzoneKey(keyInfo.getVolumeName(), keyInfo.getBucketName(),
```
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMDirectoriesPurgeResponseWithFSO.java:136]: For FSO, here we are using `/volumeId/bucketId/parentObjId/fileName/objectId` but`OMKeyDeleteResponseWithFSO` still uses `/volumeName/bucketName/keyName/objectId`
https://github.com/apache/ozone/blob/5eead92666b374c4bb912f332488acb23f9a81df/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMKeyDeleteResponseWithFSO.java#L92-L102

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java [hunk 2]
```diff
@@ -115,15 +114,15 @@ protected void addDeletionToBatch(
...
-   * @param fullKeyName (format: vol/buck/key)
+   * @param deleteKeyName (format: <keyname>/keyObjectId)
    * @param omKeyInfo
    * @throws IOException
    */
   protected void addDeletionToBatch(
       OMMetadataManager omMetadataManager,
       BatchOperation batchOperation,
       Table<String, ?> fromTable,
-      String keyName, String fullKeyName,
```
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:126]: The `fullKeyName` passed here is different still `/volumeName/bucketName/keyName/objectId` please check in other places as well.

**Review by aswinshakil — COMMENTED** [review]
With the current change the `deletedTable` will have two types of entries, please correct me if I'm wrong.
non-FSO: `/volumeName/bucketName/keyName/objectId -> RepeatedOmKeyInfo`
FSO: `/volumeId/bucketId/parentObjId/fileName/objectId -> RepeatedOmKeyInfo`

If so please update here as well, https://github.com/apache/ozone/blob/dd003040a41def491e8de003ef8539ce40854972/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L175

**Comment by prashantpogde** [comment]
This is a good find :) We need to handle this with objectID based check, the same way that we are dealing with Snapshot deletion and checking for the presence of the exact same key in the previous snapshot.

**Comment by smengcl** [comment]
Hi @sumitagrawl , @aswinshakil could help you on fixing the snapshot GC splitting keys logic (which seems to be the only part where it could go wrong). Aswin might just push the fix into this PR to make sure this PR by itself won't break that part of the logic.

**Comment by sumitagrawl** [comment]
@szetszwo as requested, This PR is splitted into 2 PR 
- HDDS-8462. Key in deleteTable uniqueness -- Raised
- HDDS-8463. S3 key uniqueness in deletedTable -- To be done after key is done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 3]
```diff
@@ -473,15 +471,7 @@ public static RepeatedOmKeyInfo prepareKeyForDelete(OmKeyInfo keyInfo,
     // Set the updateID
     keyInfo.setUpdateID(trxnLogIndex, isRatisEnabled);
 
-    if (repeatedOmKeyInfo == null) {
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:476]: @aswinshakil Thanks for review, I have created another JIRA to handle this specific issue as may need more change to have unique key, https://issues.apache.org/jira/browse/HDDS-8467

**Comment by sumitagrawl** [comment]
> With the current change the `deletedTable` will have two types of entries, please correct me if I'm wrong. non-FSO: `/volumeName/bucketName/keyName/objectId -> RepeatedOmKeyInfo` FSO: `/volumeId/bucketId/parentObjId/fileName/objectId -> RepeatedOmKeyInfo`
> 
> If so please update here as well,
> 
> https://github.com/apache/ozone/blob/dd003040a41def491e8de003ef8539ce40854972/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L175

I am keeping the old behaviour now just adding objectID, keeping /voumeName/bucketName/keyName/<objectID>
The changes is done in PR splitted now, HDDS-8462

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java (hunk 2 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:126]: @aswinshakil Its verified and handled to have /volumeName/bucketName/keyName/objectId, Its done as part of below PR splitted from this, HDDS-8462

**Pull request closed without merging**